### PR TITLE
Add NAN handling to convert() needed for some prefix routines with integer outputs.

### PR DESCRIPTION
**Pull request by rkarim2**

### CODE REVIEW

File: cunumeric/deferred.py [hunk 1]
```diff
@@ -1111,7 +1112,12 @@ def swapaxes(self, axis1: int, axis2: int) -> DeferredArray:
 
     # Convert the source array to the destination array
     @auto_convert([1])
-    def convert(self, rhs: Any, warn: bool = True) -> None:
+    def convert(
+        self,
+        rhs: Any,
+        warn: bool = True,
+        nan_op: Optional[int] = None,
```
Comment (magnatelee) [inline cunumeric/deferred.py:1119]: I think you can change this to `nan_op: CovertCode = ConvertCode.NOOP`

File: cunumeric/deferred.py [hunk 2]
```diff
@@ -1125,13 +1131,15 @@ def convert(self, rhs: Any, warn: bool = True) -> None:
                 category=UserWarning,
             )
 
+        if nan_op is None:
```
Comment (magnatelee) [inline cunumeric/deferred.py:1134]: this if statement becomes unnecessary if you make the edit suggested above.

File: cunumeric/eager.py [hunk 3]
```diff
@@ -485,21 +486,49 @@ def swapaxes(self, axis1: int, axis2: int) -> NumPyThunk:
         self.children.append(result)
         return result
 
-    def convert(self, rhs: Any, warn: bool = True) -> None:
+    def convert(
+        self,
+        rhs: Any,
+        warn: bool = True,
+        nan_op: Optional[int] = None,
```
Comment (magnatelee) [inline cunumeric/eager.py:493]: you can make a similar change here to simplify the code.

File: cunumeric/eager.py [hunk 4]
```diff
@@ -485,21 +486,49 @@ def swapaxes(self, axis1: int, axis2: int) -> NumPyThunk:
...
+                else:
+                    self.array.fill(rhs.array.item())
             else:
                 if (
                     rhs.array.dtype.kind == "c"
                     and self.array.dtype.kind != "c"
                 ):
-                    self.array[:] = rhs.array.real
+                    if nan_op is ConvertCode.SUM:
+                        self.array[:] = np.select(
```
Comment (magnatelee) [inline cunumeric/eager.py:512]: you wouldn't have to wrap the arguments with brackets if you used `np.where`.

File: src/cunumeric/unary/convert_template.inl [hunk 5]
```diff
@@ -67,21 +71,31 @@ struct ConvertImpl {
...
 };
 
 template <VariantKind KIND>
 static void convert_template(TaskContext& context)
 {
-  ConvertArgs args{context.outputs()[0], context.inputs()[0]};
-  type_dispatch(args.in.code(), SourceTypeDispatch<KIND>{}, args);
+  ConvertArgs args{
+    context.outputs()[0], context.inputs()[0], context.scalars()[0].value<ConvertCode>()};
+  op_dispatch(args.nan_op, ConvertDispatch<KIND>{}, args);
```
Comment (magnatelee) [inline src/cunumeric/unary/convert_template.inl:98]: Doing this dispatch upfront triples the number of template instantiations, even though the dispatch on `nan_op` is unnecessary when the source has an integer type (and please remember there are more integer types than floating point types and complex types combined). A more desirable implementation would be to instantiate a special conversion logic only for pairs of types that need it. You can express those pairs using a template like this:
```
template <LegateTypeCode SRC_TYPE, LegateTypeCode DST_TYPE>
using needs_dispatch_on_nan_op =
  (legate::is_floating_type(SRC_TYPE)::value || legate::is_complex_type(SRC_TYPE)::value) 
  && legate::is_integer_type(DST_TYPE)::value;
```
Then you move the dispatch on `nan_op` to the innermost template and do it only when `needs_dispatch_on_nan_op` is true.

**Review by magnatelee — CHANGES_REQUESTED** [review]

File: cunumeric/eager.py (hunk 4 above)
Comment (rkarim2) [reply cunumeric/eager.py]: 661ee85e10d7543c346b087a05851e3326a05700

File: cunumeric/eager.py (hunk 3 above)
Comment (rkarim2) [reply cunumeric/eager.py]: bd361b6dc3a341df6a95f957f32321f70827ad38

File: cunumeric/deferred.py (hunk 2 above)
Comment (rkarim2) [reply cunumeric/deferred.py]: bd361b6dc3a341df6a95f957f32321f70827ad38

File: cunumeric/deferred.py (hunk 1 above)
Comment (rkarim2) [reply cunumeric/deferred.py]: bd361b6dc3a341df6a95f957f32321f70827ad38

File: src/cunumeric/unary/convert_template.inl (hunk 5 above)
Comment (rkarim2) [reply src/cunumeric/unary/convert_template.inl]: Resolved in multiple commits, completed fix in d20450a67aca509aa04d3e184b9e1955f44f6cd1

**Review by magnatelee — APPROVED** [review]

**Comment by magnatelee** [comment]
LGTM. feel free to merge it once you fix the merge conflict

**Pull request merged**